### PR TITLE
Add a fuzzer for COLRv1 fonts API

### PR DESCRIPTION
**Pull request by drott**
Add a corpus and glyph traversal for COLRv1 fonts. Iterate over the
glyphs in the font and if a COLRv1 glyph is found, traverse its drawing
graph, retrieving gradients and their positions, transform data and
color line data.

Fixes: https://gitlab.freedesktop.org/freetype/freetype/-/issues/1030

### CODE REVIEW

**Comment by drott** [comment]
@cherusker, it'd be pretty helpful if you could start reviewing this. There might be a set of coding style issues - though I tried to fit in. 

I am still planning to split the COLRv1 fonts in the corpus into smaller files and reduce the size in the options file. I'll do that in a follow-up. I thought I'd upload already so we can start discussing the code.

**Comment by cherusker** [comment]
@drott if it's fine with you I'll take a look at this at the weekend! 😊

**Comment by drott** [comment]
Sure, whenever you have time. If you have any ad-hoc comments earlier, feel free to drop those already. I'll update the corpus today.

**Comment by drott** [comment]
Friendly ping @cherusker.

File: fuzzing/src/targets/font-drivers/colrv1.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#include "targets/font-drivers/colrv1.h"
+
+#include <utility> // std::move
+
+#include "iterators/faceloaditerator.h"
+#include "iterators/faceprepiterator-bitmaps.h"
+#include "iterators/faceprepiterator-outlines.h"
+#include "visitors/facevisitor-colrv1.h"
+
+  freetype::ColrV1FuzzTarget::
```
Comment (cherusker) [inline fuzzing/src/targets/font-drivers/colrv1.cpp:25]: for consistency's sake: let's please have two empty lines between `#include`s and class definitions

File: fuzzing/src/targets/font-drivers/colrv1.cpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  ColrV1FuzzTarget()
+  {
+    auto  fli = freetype::make_unique<FaceLoadIterator>();
+
```
Comment (cherusker) [inline fuzzing/src/targets/font-drivers/colrv1.cpp:29]: for consistency's sake: let's please have two empty lines before the C90-like variable declarations and the first line of code within functions

File: fuzzing/src/visitors/facevisitor-colrv1.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    : public FaceVisitor
+  {
+  public:
+
+
+    void
+    run( Unique_FT_Face  face )
+    override;
+
+  private:
```
Comment (cherusker) [inline fuzzing/src/visitors/facevisitor-colrv1.h:36]: for consistency's sake: let's please have two empty lines here

File: fuzzing/src/visitors/facevisitor-colrv1.cpp [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+#include "utils/logging.h"
+
+#include <ft2build.h>
+#include FT_FREETYPE_H
+
+namespace {
+
+bool colrv1_start_glyph( const FT_Face& ft_face,
+                         uint16_t glyph_id,
+                         FT_Color_Root_Transform root_transform );
```
Comment (cherusker) [inline fuzzing/src/visitors/facevisitor-colrv1.cpp:29]: this applies to all function signatures: the way that FreeType uses that in the C codebase (and what I tried to mimic) is the following layout (e.g.):

```cpp
bool colrv1_start_glyph( const FT_Face&           ft_face,
                         uint16_t                 glyph_id,
                         FT_Color_Root_Transform  root_transform );
```

- let's left-align all variable names
- let's have two spaces between the type + the name

File: fuzzing/src/visitors/facevisitor-colrv1.cpp [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+void iterate_color_stops ( FT_Face face, FT_ColorStopIterator* color_stop_iterator ) {
```
Comment (cherusker) [inline fuzzing/src/visitors/facevisitor-colrv1.cpp:31]: let's please make those private class members 😊

**Review by cherusker — COMMENTED** [review]
thanks a lot! sorry it took that long; started looking through it, got pulled away, continued a day later + so on 😇 

First of all: thanks a lot for looking through all this + finding a way of working with the framework 😊 I did leave a couple of comments but they're all just code style more or less; logic wise it looks very good but I'll give it another thorough read.

As for styling the `*.cpp` file; do you mind taking a look at https://github.com/freetype/freetype2-testing/blob/master/fuzzing/src/visitors/facevisitor-multiplemasters.cpp - esp. in terms of linebreaking + indentation. Unfortunately we don't have a `clang-format` for this project yet 😞 Happy to chat specifics if there are any doubts; the main thing is probably the rather unusual use of double empty lines (to put more emphasis on coherent blocks of code like functions) + double spaces (between variable types + their names). Other than it's mainly just `{` on a new line and indentation with spaces to align statements that span across multiple lines.

Thanks again! 😊

File: fuzzing/README.md [hunk 6]
```diff
@@ -374,6 +374,19 @@ https://www.freetype.org/freetype2/docs/reference/ft2-bitmap_handling.html
...
+
+Not yet released, see documentation in [freetype.h](https://gitlab.freedesktop.org/freetype/freetype/-/blob/master/include/freetype/freetype.h#L4137)
+
+|    | Function                   | Fuzzed             |
+| -- | -------------------------- | :----------------: |
+|    | `FT_Get_Color_Glyph_Paint` | :heavy_check_mark: |
+|    | `FT_Get_Paint_Layers`      | :heavy_check_mark: |
+|    | `FT_Get_Paint_Layers`      | :heavy_check_mark: |
+|    | `FT_Get_Colorline_Stops`   | :heavy_check_mark: |
+|    | `FT_Get_Paint`             | :heavy_check_mark: |
```
[the remark this answers is not in the document]
Comment (lemzwerg) [reply fuzzing/README.md:387]: @cherusker: Right now, all of the COLR v2 stuff is part of the 'Glyph Layer Management' section, having the following order (including structures and the like).
```
FT_LayerIterator
FT_Get_Color_Glyph_Layer
FT_PaintFormat
FT_ColorStopIterator
FT_ColorIndex
FT_ColorStop
FT_PaintExtend
FT_ColorLine
FT_Affine23
FT_Composite_Mode
FT_OpaquePaint
FT_PaintColrLayers
FT_PaintSolid
FT_PaintLinearGradient
FT_PaintRadialGradient
FT_PaintSweepGradient
FT_PaintGlyph
FT_PaintColrGlyph
FT_PaintTransformed
FT_PaintTranslate
FT_PaintRotate
FT_PaintSkew
FT_PaintComposite
FT_COLR_Paint
FT_Color_Root_Transform
FT_Get_Color_Glyph_Paint
FT_Get_Paint_Layers
FT_Get_Colorline_Stops
FT_Get_Paint
```
I'm open to suggestions to improve that, if you feel this might be useful.

File: fuzzing/src/visitors/facevisitor-colrv1.cpp (hunk 4 above)
Comment (drott) [reply fuzzing/src/visitors/facevisitor-colrv1.cpp]: Done, hope I didn't miss any occasions.

File: fuzzing/src/visitors/facevisitor-colrv1.h (hunk 3 above)
Comment (drott) [reply fuzzing/src/visitors/facevisitor-colrv1.h]: Done.

File: fuzzing/src/targets/font-drivers/colrv1.cpp (hunk 2 above)
Comment (drott) [reply fuzzing/src/targets/font-drivers/colrv1.cpp:30]: Done.

File: fuzzing/src/targets/font-drivers/colrv1.cpp (hunk 1 above)
Comment (drott) [reply fuzzing/src/targets/font-drivers/colrv1.cpp]: Done.

File: fuzzing/src/visitors/facevisitor-colrv1.h [hunk 7]
```diff
@@ -0,0 +1,44 @@
+// facevisitor-colrv1.h
+//
+//   Access and traverse COLRv1 glyph formation.
+//
+// Copyright 2018-2019 by
+// Armin Hasitzka.
```
[the remark this answers is not in the document]
Comment (drott) [reply fuzzing/src/visitors/facevisitor-colrv1.h]: Done.

File: fuzzing/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (drott) [reply fuzzing/README.md]: Resources column added.

File: fuzzing/settings/oss-fuzz/colrv1.options [hunk 8]
```diff
@@ -0,0 +1,2 @@
+[libfuzzer]
+max_len = 120000
```
[the remark this answers is not in the document]
Comment (drott) [reply fuzzing/settings/oss-fuzz/colrv1.options:2]: `noto_complex_glyphs_colrv1_0.ttf` is 104208, I choose this size to be a little larger than the largest of the fonts from the corpus for colrv1.

File: fuzzing/src/visitors/facevisitor-colrv1.cpp (hunk 5 above)
Comment (drott) [reply fuzzing/src/visitors/facevisitor-colrv1.cpp]: May I ask what's the reasoning behind this suggestion? These are pure helper functions implementing the recursive traversal of the COLRv1 glyph graph, they're not accessing class members of `FaceVisitorColrV1` and are local to this `FaceVisitor`. FWIW the Google C++ style guide [suggests prefering internal linkage](https://google.github.io/styleguide/cppguide.html#Internal_Linkage)). Seems to me it'd make sense to keep them in this anonymous namespace.

**Comment by drott** [comment]
> As for styling the `*.cpp` file; do you mind taking a look at https://github.com/freetype/freetype2-testing/blob/master/fuzzing/src/visitors/facevisitor-multiplemasters.cpp - esp. in terms of linebreaking + indentation. Unfortunately we don't have a `clang-format` for this project yet 😞 Happy to chat specifics if there are any doubts; the main thing is probably the rather unusual use of double empty lines (to put more emphasis on coherent blocks of code like functions) + double spaces (between variable types + their names). Other than it's mainly just `{` on a new line and indentation with spaces to align statements that span across multiple lines.

Tried to incorporate that. Thanks for the instructions. +1 for having automatic formatting, coming from Blink the lack of automatic formatting and no clear place for where to find coding conventions is a bit of an obstacle to contributing to FreeType and freetype2-testing. 

Thanks for your review, Armin - hope these changes address your comments.

**Comment by drott** [comment]
What are your thoughts, Armin - can we proceed here?

**Comment by cherusker** [comment]
> What are your thoughts, Armin - can we proceed here?

Thanks for looking into this! 😊 I will go over this manually at an evening next week and suggest a patch in terms of code style (spacing, moving C-style static functions into classes' private sections, moving `{` to new lines, double spaces between variable types + names, etc)

Sorry for the delays of my answers btw; unfortunately, I'm very busy with my current dayjob and this weekend is fully booked already. I will, however, find an evening in the next 5 days for this!

Thanks for haning in there!

File: fuzzing/src/visitors/facevisitor-colrv1.cpp (hunk 5 above)
Comment (cherusker) [reply fuzzing/src/visitors/facevisitor-colrv1.cpp]: oh sorry missed this comment before I answered below; from my point of view converting free functions to static class members helps with "finding" their declarations e.g. in automated documentations or simply by browsing class definitions.

If I read the internal linkage paragraph correctly, it mainly suggests to not put declarations of module-private free functions in header files for public access (which I agree with completley). The paragraph below "Nonmember, Static Member, and Global Functions" leaves a bit more room about whether to use free functions or static members. In this case, "Static methods of a class should generally be closely related to instances of the class" seems to be fulfilled from my point of view.

In C++ projects I have worked on so far (commerical and private) I've always found that static members keep the code cleaner + easier to follow (one place to look for declarations, one place to look for definitions) but not too fussed here. Happy to leave it as is if you prefer that 😊

File: fuzzing/settings/oss-fuzz/colrv1.options (hunk 8 above)
[the remark this answers is not in the document]
Comment (cherusker) [reply fuzzing/settings/oss-fuzz/colrv1.options:2]: fair enough 😊

File: fuzzing/src/visitors/facevisitor-colrv1.cpp (hunk 5 above)
Comment (drott) [reply fuzzing/src/visitors/facevisitor-colrv1.cpp]: > If I read the internal linkage paragraph correctly, it mainly suggests to not put declarations of module-private free functions in header files for public access (which I agree with completley). The paragraph below "Nonmember, Static Member, and Global Functions" leaves a bit more room about whether to use free functions or static members. In this case, "Static methods of a class should generally be closely related to instances of the class" seems to be fulfilled from my point of view.

I read the main clause 

> When definitions in a .cc file do not need to be referenced outside that file, give them internal linkage by placing them in an unnamed namespace or declaring them static. 

to really mainly focus on linkage and not about public access / interface questions. I recommends to give only internal linkage to anything that's not referenced/used outside the implementation file - which in our case applies to the COLRv1 traversal function implementations. 

That's also common practice in Blink for file-local helper functions etc. - so if you're okay with that I am happy to leave them there. I don't have too strong of an opinion on it either. I was mainly curious what's the background of the request.

**Review by cherusker — COMMENTED** [review]
@drott I applied the FT code style to the best of my knowledge; do you mind applying https://gist.github.com/cherusker/a041262d0596fb0003ec88cbcaca70fd to your PR + quickly double check that I haven't introduced any bugs (had to move / split a definition or two) ... also, happy to chat details if you have questions and/or complaints - also hope I haven't missed anything ... this is still a manual approach unfortunately 😞

File: fuzzing/src/visitors/facevisitor-colrv1.cpp (hunk 5 above)
Comment (cherusker) [reply fuzzing/src/visitors/facevisitor-colrv1.cpp]: thanks for the extened conversation on this 😊 all the times I've built CEF recently, I haven't looked into the code yet ... do you happen to know the rough % of static helpers vs. class member functions in Blink? would be interesting to know whether building (esp. linking) would take even longer without strong focus on internal linking ... only downside I think to see is code clarity ("finding" relevant pieces directly in the code or in automated docs) ... might be wrong though + could just be a matter of getting used to it. an interesting thought regardless; will keep that in mind for future projects!

TL;DR let's leave it as is then 😊 only suggestion I had would be to move the `constexpr` definition to the top to roughly follow the pattern of "variables first, then functions" (suggeseted that in my gist)

File: fuzzing/src/visitors/facevisitor-colrv1.cpp [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+    {
+      LOG ( INFO ) << "PaintSkew," <<
+          " center.x " << colrv1_paint.u.skew.center_x <<
+          " center.y " << colrv1_paint.u.skew.center_y <<
+          " x_skew_angle " << colrv1_paint.u.skew.x_skew_angle <<
+          " y_skew_angle " << colrv1_paint.u.skew.y_skew_angle;
+      break;
+    }
+    default:
+      break;
```
[the remark this answers is not in the document]
Comment (drott) [reply fuzzing/src/visitors/facevisitor-colrv1.cpp]: Not a bad idea, this actually uncovered that I was missing `TRANSLATE` cases. Added those, made the `default` clauses explicit here and in `colrv1_start_glyph`. Added additional logging to cover `PaintGlyphs`, `PaintColrGlyph`, `PaintColorLayers` there too. Those types are not outputting anything in the draw function.

**Comment by drott** [comment]
> @drott I applied the FT code style to the best of my knowledge; do you mind applying https://gist.github.com/cherusker/a041262d0596fb0003ec88cbcaca70fd to your PR + quickly double check that I haven't introduced any bugs (had to move / split a definition or two) ... also, happy to chat details if you have questions and/or complaints - also hope I haven't missed anything ... this is still a manual approach unfortunately 😞

Done, thanks for doing that, again trying to memorize a couple of additional rules. (The merge tool was struggling to show this diff right, but I see no issues, everything's in place). 

Hope that's it then - thanks for the review.

**Comment by cherusker** [comment]
@drott sounds + looks good to me! 😊 just to quickly confirm; are you happy with the PR from your end? in this case I'll merge it for you 😊

also, sorry that this took forever; if you plan to make more contributions like this down the road I'd love to streamline the whole process a lot more!

💛

File: fuzzing/src/visitors/facevisitor-colrv1.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (cherusker) [reply fuzzing/src/visitors/facevisitor-colrv1.cpp]: 😃

**Review by cherusker — APPROVED** [review]

**Comment by drott** [comment]
Thanks, yes, please go ahead and merge. - If you could assign commit rights to my account on this repo, I could do that myself.

**Comment by cherusker** [comment]
oh sorry one last thing I just saw -- you did bump `freetype2` from `851684` to `2149b5` ... did you mean to do this? (it shouldn't generally matter as the fuzzers pull latest `master` anyways but just curious 😊

**Comment by drott** [comment]
Ah, that made it into the commit - I think I had to locally update the submodules otherwise I saw git complaining. I can remove that from the change, one sec.

**Comment by drott** [comment]
Okay, done - feel free to merge.
